### PR TITLE
Feature: Seek-to-live button with related events & states

### PR DESCRIPTION
**Pull request by heff**
Modeled after Mux Player's implementation.

### `<media-live-button>`
image

Indicate the video is currently live and click to seek to live. Includes slots for alt icons and test, and css variables for changing colors etc.

### mediaseektoliverequest event
Event that when handled will seek to the end of the seekable window

### media-time-is-live attribute
State for when the current time is near enough to the live edge to be considered "live"

### Todo

- [x] Docs/Examples
- [x] Finish MEDIA_TIME_IS_LIVE state
- [x] Optional auto seek to live on unpause

### CODE REVIEW

File: src/js/media-controller.js
Comment (cjpillsbury) [inline src/js/media-controller.js:1018]: idea(non-blocking): This is the sort of thing that could be more reliably computed by a(n extended) slotted media element based on spec-compliant/recommended behaviors on refetches, segment sizes, etc. etc. Would be a good followup to spike on.

File: src/js/media-live-button.js
Comment (cjpillsbury) [inline src/js/media-live-button.js:58]: suggestion(non-blocking): We should add `MEDIA_STREAM_TYPE` to the list of observed attributes to make styling cases easier.

File: src/js/media-controller.js
Comment (heff) [reply src/js/media-controller.js:1018]: Yeah, I agree. That'd be a good way to set your latency preferences on the media element specifically and have it impact the right things in the UI, without latency needing to be a bigger concept (at least yet).

Start a media ui extension issue?

File: src/js/media-live-button.js
Comment (heff) [reply src/js/media-live-button.js:58]: 👍 done

**Comment by heff** [comment]
Notes from IRL conversation:
* Pin the progress bar to the right when in the live window
* Always seek to the live edge during unpause
* Turn off red indicator when unpaused, at least until we know the difference between seekable and non-seekable live streams

**Comment by heff** [comment]
Office hours request: Option to turn on/off seek-to-live on unpause

**Comment by heff** [comment]
- [x] Docs completed
- [x] Optional seek to live on unpause added. Attr is `autoseektolive="off"`

File: docs/src/pages/en/media-live-button.md
Comment (dylanjha) [inline docs/src/pages/en/media-live-button.md:115]: docs look great, but needs to be added to the side nav in `config.ts`:

https://github.com/muxinc/media-chrome/blob/main/docs/src/config.ts#L65

**Review by dylanjha — CHANGES_REQUESTED** [review]

File: docs/src/pages/en/media-live-button.md
Comment (gkatsev) [inline docs/src/pages/en/media-live-button.md:115]: thought:
- `--media-live-indicator-icon-color` for the default/non-live color
- `--media-live-indicator-live-icon-color` for the live color

File: src/js/media-controller.js
Comment (gkatsev) [inline src/js/media-controller.js:98]: `autoseektolive=off` doesn't feel very htmly also increases inconsistency between attributes in media-chrome. That said, don't have any alternatives off the top of my head currently.

File: src/js/media-live-button.js
Comment (gkatsev) [inline src/js/media-live-button.js:8]: nit: unused
```suggestion
```

File: src/js/media-live-button.js
Comment (gkatsev) [inline src/js/media-live-button.js:62]: nit: unused `e`
```suggestion
  handleClick() {
```

File: src/js/media-controller.js
Comment (luwes) [reply src/js/media-controller.js:98]: yes for a boolean attribute we should try to not use `off`, `false` as value.
best to inverse it then.

`noseektolive`, `disableliveseek`, `noplaylive`

Comment (heff) [reply src/js/media-controller.js:98]: Yeah naming this one was hard for me. Open to other ideas but I feel good enough about where it landed. It's also flexible incase we wanted to capture other options with it, like maybe one that tries to keep the playhead at the live edge even without unpausing.

> increases inconsistency between attributes in media-chrome

How so?

File: docs/src/pages/en/media-live-button.md
Comment (cjpillsbury) [inline docs/src/pages/en/media-live-button.md:98]:
```suggestion
_None_
```
Currently, we only document attributes intended to be set by an implementor for customization/configuration (and not attributes set by the Media Controller). See, e.g. [media-play-button.md](https://github.com/muxinc/media-chrome/blob/main/docs/src/pages/en/media-play-button.md). If we'd like to document both (ideally with some differentiation between those two with an eye towards devex), we should plan this and do it as a single effort.

**Review by cjpillsbury — CHANGES_REQUESTED** [review]

File: docs/src/pages/en/media-live-button.md
Comment (heff) [reply docs/src/pages/en/media-live-button.md:115]: I see how you get there, and feel the awkwardness in it. Feels a little weird to have "live" in there twice, and I expect "live indicator color" to refer to the red not the "off" state.

I changed it to:
* `--media-live-indicator-color` - squarely focused on the color of live and re-useable
* `--media-live-button-icon-color` - because what it's really doing is overriding the default `--media-icon-color`

Make sense?

File: docs/src/pages/en/media-live-button.md
Comment (heff) [reply docs/src/pages/en/media-live-button.md:98]: Yeah, I saw that, and I disagree with only documenting the non-media-controller attributes, based on the goal of [element independence](https://github.com/muxinc/media-chrome/blob/main/ARCHITECTURE.md#element-independence). I also don't love requiring a big project to ship it all at once since the temporary inconsistency doesn't really hurt anyone and it's no small effort to document all the detail per attr per element. But I agree in that I think we should make it a goal of 1.0 to document those attrs everywhere so the inconsistency has a clear end date, and in that we can create some better visual separation between them.

File: docs/src/pages/en/media-live-button.md
Comment (heff) [reply docs/src/pages/en/media-live-button.md]: Added. Good now?

**Review by dylanjha — APPROVED** [review]

File: docs/src/config.ts
Comment (gkatsev) [inline docs/src/config.ts:82]: Can this be moved above `Loading Inidcator` to keep this list in alphabetical order?

Comment (heff) [reply docs/src/config.ts:82]: Ah, missed that. Yes. Thanks.

File: src/js/media-controller.js
Comment (luwes) [reply src/js/media-controller.js:98]: `autohide="2"`
`gestures-disabled` why not `gestures="off"` then?
`nohotkeys` => better remove this one and add an `hotkeys="off"` option
`hotkeys="noarrowleft noarrowright"` 
`default-stream-type`
`audio`

I agree that the uniformness is starting to worsen between these handful of attributes.

we should decide 
- dashes vs no dashes
- how should a disabled attribute be. if it's `feature="off"` fine with me but best to keep it consistent

File: docs/src/pages/en/media-live-button.md
Comment (cjpillsbury) [reply docs/src/pages/en/media-live-button.md:101]: I do worry that having this one place where we document the "internal" attributes will be confusing in the interim and increases the chance of mistakes when we **_do_** update the docs to include both. I'd still recommend that we don't document currently (there's already some implicit documentation of those attributes above in terms how how they update the visual state), but I won't treat this as a blocker.

File: src/js/media-controller.js
Comment (cjpillsbury) [reply src/js/media-controller.js:98]: I'd agree with @gkatsev and @luwes here. For an attribute currently named `autoseektolive`, I'm having a hard time imagining expanding values beyond a `boolean` (though that may just be a failure of imagination). As such, we might as well stick with both our convention and the common convention for boolean attributes in HTML: if the attribute is present (typically with a value of `""`), that implies `true`. Otherwise (aka it's not present), that implies `false`.

**Review by gkatsev — APPROVED** [review]

File: src/js/media-controller.js
Comment (cjpillsbury) [reply src/js/media-controller.js]: @heff Yeah, good call re: media ui extension.

File: docs/src/pages/en/media-live-button.md
Comment (cjpillsbury) [reply docs/src/pages/en/media-live-button.md]: This might be a good thing for our next media chrome weekly meeting agenda. A thought I'm having is:
Should this be two separate variables at all? Should we instead just assume folks can currently use attribute selectors and potentially eventually use [Custom States](https://developer.mozilla.org/en-US/docs/Web/API/CustomStateSet) (which we keep running into candidate use cases for)?

File: src/js/media-controller.js
Comment (heff) [reply src/js/media-controller.js:98]: Yeah I see the reason here, just trying to get this thing shipped already. 😆 Happy to give up on a multi-value approach if no one sees opportunity for expansion.

`<form>` at least has `novalidate`, so there's some precedence for no*, and I can get behind that. I think my favorite might be `noautoseektolive` (or dashed). `noseektolive` would be less verbose, but that kind of sounds like we're disabling the button from seeking to live too (which is why I added `auto` in the first place).

I can't think of any good option to reverse the relationship so it's not a negative. `manualseektolive`, `maintainliveplayheadposition`, `makeuserclicklivebuttontoseektolive`. :)

File: src/js/media-live-button.js
Comment (heff) [reply src/js/media-live-button.js:38]: This is a good idea. Just tried this and had the same result of mismatched spacing. There's something weird happening between slotted elements and default slots that I can't quite figure out. Going to ship as is but feel free to try your hand at it if you're interested enough. If I'm not just missing something dumb I'm sure we'll run into it again.

File: docs/src/pages/en/media-live-button.md
Comment (heff) [reply docs/src/pages/en/media-live-button.md]: You can't use attribute selectors to target the indicator color specifically of the built-in indicator. You'd have to override the default slot to use attribute selectors. Or we'd have to expose it as a part. I could be fine with the latter, but I'm not sure how that works if the part is also a slot, and then implications of that for skins/themes.

CustomStateSet is the dream. Do we have a sense for how long we're going to be waiting for those?

**Pull request merged**
